### PR TITLE
engine: wire context into methods that invoke Docker API call

### PR DESCRIPTION
**Pull request by haikuoliu**
This pr is to fix issue #1212.

### Summary

Pass the context into the Docker API call instead of creating context inside Docker API call. 

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/docker_client.go [hunk 1]
```diff
@@ -125,31 +125,32 @@ type DockerClient interface {
...
-	CreateContainer(*docker.Config, *docker.HostConfig, string, time.Duration) DockerContainerMetadata
+	// and a context should be provided for the request.
+	CreateContainer(*docker.Config, *docker.HostConfig, string, time.Duration, context.Context) DockerContainerMetadata
 
-	// StartContainer starts the container identified by the name provided. A timeout value should be provided for the
-	// request.
-	StartContainer(string, time.Duration) DockerContainerMetadata
+	// StartContainer starts the container identified by the name provided. A timeout value and a context should be
+	// provided for the request.
+	StartContainer(string, time.Duration, context.Context) DockerContainerMetadata
```
Comment (petderek) [inline agent/engine/docker_client.go]: Actually, this is all repetitive as far as I can tell. Lets get rid of time.Duration in all of the signatures.

File: agent/engine/docker_client_test.go [hunk 2]
```diff
@@ -438,20 +440,24 @@ func TestStartContainerTimeout(t *testing.T) {
...
 	defer done()
 
-	testDone := make(chan struct{})
 	wait := &sync.WaitGroup{}
 	wait.Add(1)
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockDone := make(chan struct{})
+	mockContext := NewMockContext(mockCtrl)
+	mockContext.EXPECT().Done().Return(mockDone).AnyTimes()
```
Comment (richardpen) [inline agent/engine/docker_client_test.go]: Why those are `AnyTimes`, shouldn't this be called just once or twice?

File: agent/engine/docker_client.go [hunk 3]
```diff
@@ -474,15 +475,8 @@ func (dg *dockerGoClient) getAuthdata(image string, authData *api.RegistryAuthen
...
-	// Create a context that times out after the 'timeout' duration
-	// This is defined by the const 'createContainerTimeout'. Injecting the 'timeout'
-	// makes it easier to write tests.
-	// Eventually, the context should be initialized from a parent root context
-	// instead of TODO.
-	ctx, cancel := context.WithTimeout(context.TODO(), timeout)
-	defer cancel()
-
+	timeout time.Duration,
+	ctx context.Context) DockerContainerMetadata {
```
[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/engine/docker_client.go]: Sounds reasonable, so can I just return the timeout error here without checking the error type from the context?

File: agent/engine/docker_client.go [hunk 4]
```diff
@@ -474,15 +475,8 @@ func (dg *dockerGoClient) getAuthdata(image string, authData *api.RegistryAuthen
...
-	timeout time.Duration) DockerContainerMetadata {
-	// Create a context that times out after the 'timeout' duration
-	// This is defined by the const 'createContainerTimeout'. Injecting the 'timeout'
-	// makes it easier to write tests.
-	// Eventually, the context should be initialized from a parent root context
-	// instead of TODO.
-	ctx, cancel := context.WithTimeout(context.TODO(), timeout)
-	defer cancel()
-
+	timeout time.Duration,
```
[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/engine/docker_client.go]: When returning a timeout error, the timeout value is needed, how to get this value?

File: agent/engine/docker_client.go (hunk 1 above)
Comment (haikuoliu) [reply agent/engine/docker_client.go]: Please see the comments above.

File: agent/engine/docker_client_test.go (hunk 2 above)
Comment (haikuoliu) [reply agent/engine/docker_client_test.go]: I think it should be called twice, because the `CreateContainer` will also call `InspectContainer` where a new child context will be generated, but test using mock context will be deleted anyway.

File: agent/engine/docker_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (petderek) [reply agent/engine/docker_client.go]: If we really want to print the timeout as part of the error, we should assemble it in the caller (which already has both of these pieces of information).

File: agent/engine/docker_client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (petderek) [reply agent/engine/docker_client.go]: Precisely!

File: agent/engine/docker_client_test.go (hunk 2 above)
Comment (petderek) [reply agent/engine/docker_client_test.go]: My opinion: The number of times context is invoked is an implementation detail and not part of tested function's contract.  IE, its not something we would want the test to fail for.

File: agent/engine/docker_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/engine/docker_client.go]: I think logging the timeout is important, because most of the timeout values are not defined by users, so they have no idea how long they've waited for the API call in the most of the time. 

But assembling the error from the caller requires at least three lines of the same code, and this same piece of code will be spread wherever the Dock client API is called, and will require more code change, what do you think?

**Comment by haikuoliu** [comment]
Situation when the go routine will not return using mock context:
Let's take the `TestCreateContainerTimeout` for example. 

In `CreateContainer` (call it `goroutine1`) it will create another go routine (call it `goroutine2`) to execute `createContainer`, and `goroutine2` will execute the real Docker API call, the one that we've mocked out. So instead of doing the real Docker API call, `goroutine2` will close the `mockDone` channel, and wait until the `wait.Done()` is executed. After the `mockDone` is closed, the `goroutine1` will exit and return the err as expected. Then the `wait.Done()` is executed, the mocked real Docker API call will return with the return value specified in the call chain of mock method. 

However, returned from mocked real Docker API call, `goroutine2` will continue to execute (it will not execute when using real context), and thus call `containerMetadata` and `InspectContainer`, which will result in a missing call error. But since the error happens after the test is done, the golang runtime will not give you the missing call error directly, it will give you error like 'some error happen after some test has completed' (don't remember the exact error text), so it's hard to observe this kind of bug in the test. 

The good part is I do learn a lot from this. Let me know if anything is unclear.

File: agent/engine/docker_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (petderek) [reply agent/engine/docker_client.go]: You're right. I looked at the calling class (which is embedded in the transition function map) and I don't think it makes sense to change that at this point.

[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/engine/docker_client.go]: OK, cool, I'll make the change accordingly.

**Comment by haikuoliu** [comment]
Update: use real context in the test; for docker api call, return timeout error directly when context is done

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
 ## 1.17.3-dev
+* Enhancement - Wire context into methods that invoke Docker API call to fix flaky tests [#1329](https://github.com/aws/amazon-ecs-agent/pull/1321)
```
Comment (yunhee-l) [inline CHANGELOG.md]: I don't think we normally update Changelog for internal text fixes like this. Also, 1.17.3 just got released so it shouldn't be under 1.17.3-dev.

File: agent/containermetadata/manager.go [hunk 6]
```diff
@@ -117,8 +118,10 @@ func (manager *metadataManager) Create(config *docker.Config, hostConfig *docker
 
 // Update updates the metadata file after container starts and dynamic metadata is available
 func (manager *metadataManager) Update(dockerID string, task *api.Task, containerName string) error {
+	ctx, cancel := context.WithTimeout(context.TODO(), inspectContainerTimeout)
```
Comment (aaithal) [inline agent/containermetadata/manager.go]: You should modify `Update(` to expect an arg for `context.Context` as well. The `DockerTaskEngine` struct already has a context, which it can pass in here.

File: agent/engine/docker_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaithal) [reply agent/engine/docker_client.go]: For now, lets update the documentation for all of these methods that `timeout` is only used to generate the error and shouldn't be used for anything else. We can also add a `TODO` for changing these methods to instead accept an error type or error string as a param to use when timeout error occurs.

File: agent/engine/docker_client.go [hunk 7]
```diff
@@ -493,15 +487,9 @@ func (dg *dockerGoClient) CreateContainer(config *docker.Config,
...
-		// Context has either expired or canceled. If it has timed out,
-		// send back the DockerTimeoutError
-		err := ctx.Err()
-		if err == context.DeadlineExceeded {
-			return DockerContainerMetadata{Error: &DockerTimeoutError{timeout, "created"}}
-		}
-		// Context was canceled even though there was no timeout. Send
-		// back an error.
-		return DockerContainerMetadata{Error: &CannotCreateContainerError{err}}
+		// Context has expired, send back the DockerTimeoutError
```
Comment (aaithal) [inline agent/engine/docker_client.go]: Why was this modified. I think there's a case to be made for the previous code, where we distinguish between cancelled and timed out contexts.

File: agent/engine/docker_client.go [hunk 8]
```diff
@@ -125,31 +125,32 @@ type DockerClient interface {
...
+	// StartContainer starts the container identified by the name provided. A timeout value and a context should be
+	// provided for the request.
+	StartContainer(string, time.Duration, context.Context) DockerContainerMetadata
 
-	// StopContainer stops the container identified by the name provided. A timeout value should be provided for the
-	// request.
-	StopContainer(string, time.Duration) DockerContainerMetadata
+	// StopContainer stops the container identified by the name provided. A timeout value and a context should be provided
+	// for the request.
+	StopContainer(string, time.Duration, context.Context) DockerContainerMetadata
```
Comment (aaithal) [inline agent/engine/docker_client.go]: here, and in all methods, please make sure that `context.Context` is the first parameter to all the methods (which is the convention in golang)

File: CHANGELOG.md (hunk 5 above)
Comment (haikuoliu) [reply CHANGELOG.md]: OK, I will remove it.

File: agent/engine/docker_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/engine/docker_client.go]: OK, sounds good.

File: agent/engine/docker_client.go (hunk 8 above)
Comment (haikuoliu) [reply agent/engine/docker_client.go]: OK

File: agent/containermetadata/manager.go (hunk 6 above)
Comment (haikuoliu) [reply agent/containermetadata/manager.go]: OK, will change accordingly.

File: agent/engine/docker_client.go (hunk 7 above)
Comment (haikuoliu) [reply agent/engine/docker_client.go]: As Derek said, when the caller wants to cancel the context, the caller already know the reason,  the logic to handle the timeout or cancel is nothing but report the reason of the error.

**Comment by haikuoliu** [comment]
Update: some function signatures and comments of Docker client API, change log, Update function in manager.go

File: agent/engine/docker_client_test.go [hunk 9]
```diff
@@ -361,18 +360,20 @@ func TestCreateContainerTimeout(t *testing.T) {
 	mockDocker, client, _, _, _, done := dockerClientSetup(t)
 	defer done()
 
-	warp := make(chan time.Time)
+	ctx, cancel := context.WithTimeout(context.TODO(), xContainerLongTimeout)
+	defer cancel()
```
Comment (sharanyad) [inline agent/engine/docker_client_test.go]: Since we expect the mock call to happen and we do a `cancel()` there, this `cancel()` can be removed.

File: agent/engine/docker_client.go [hunk 10]
```diff
@@ -686,33 +643,25 @@ func (dg *dockerGoClient) stopContainer(ctx context.Context, dockerID string) Do
...
 	case resp := <-response:
 		return resp
 	case <-ctx.Done():
-		err := ctx.Err()
-		// Context has either expired or canceled. If it has timed out,
-		// send back the DockerTimeoutError
-		if err == context.DeadlineExceeded {
-			return &DockerTimeoutError{removeContainerTimeout, "removing"}
-		}
-		return &CannotRemoveContainerError{err}
```
Comment (sharanyad) [inline agent/engine/docker_client.go]: Since we are not using this error type anymore, it can be removed.

Comment (sharanyad) [reply agent/engine/docker_client.go]: Giving it some thought, you can capture the error [here](https://github.com/aws/amazon-ecs-agent/blob/487e699ce931a374d5e92a2ea6ddda702a637a69/agent/engine/docker_client.go#L720) and do
```
if err != nil {
    return &CannotRemoveContainerError{err}
}
```

File: agent/engine/docker_client.go (hunk 7 above)
Comment (aaithal) [reply agent/engine/docker_client.go]: That's not strictly true. If you look at https://github.com/aws/amazon-ecs-agent/pull/1329/files#diff-b9aa12c3425c34d6f330f28407f779f6R835, you'll see that the caller (owner of the context) is not the entity that's dealing with the error. I think there's value in the way we were previously handling errors until we figure out a better alternative.

Comment (haikuoliu) [reply agent/engine/docker_client.go]: OK, cool, will change accordingly.

File: agent/engine/docker_image_manager.go [hunk 11]
```diff
@@ -345,7 +345,9 @@ func (imageManager *dockerImageManager) deleteImage(imageID string, imageState *
 		return
 	}
 	seelog.Infof("Removing Image: %s", imageID)
-	err := imageManager.client.RemoveImage(imageID, removeImageTimeout)
+	ctx, cancel := context.WithTimeout(context.Background(), removeImageTimeout)
```
Comment (aaithal) [inline agent/engine/docker_image_manager.go]: You shouldn't be using `context.Background()` here. Instead, it should be wired all the way from `performPeriodicImageCleanup()`.

File: agent/engine/docker_task_engine.go [hunk 12]
```diff
@@ -346,8 +348,9 @@ func (engine *DockerTaskEngine) synchronizeContainerStatus(container *api.Docker
 		}
 		return
 	}
-
-	currentState, metadata := engine.client.DescribeContainer(container.DockerID)
+	ctx, cancel := context.WithCancel(engine.ctx)
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: why are you doing this? Can't you just use `engine.ctx` directly?

File: agent/engine/docker_task_engine.go [hunk 13]
```diff
@@ -392,7 +395,9 @@ func (engine *DockerTaskEngine) CheckTaskState(task *api.Task) {
 		if !ok {
 			continue
 		}
-		status, metadata := engine.client.DescribeContainer(dockerContainer.DockerID)
+		ctx, cancel := context.WithCancel(engine.ctx)
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: Same question as above. Can't you just use `engine.ctx` directly?

File: agent/engine/docker_task_engine.go [hunk 14]
```diff
@@ -825,8 +830,9 @@ func (engine *DockerTaskEngine) createContainer(task *api.Task, container *api.C
 				task.Arn, container.Name, mderr)
 		}
 	}
-
-	metadata := client.CreateContainer(config, hostConfig, dockerContainerName, createContainerTimeout)
+	ctx, cancel := context.WithTimeout(engine.ctx, createContainerTimeout)
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: This is not an optimal way of doing this. All of the responsibility of creating the child context for timing out should be happening within the docker_client methods. Please change all of  these methods so that context creation happens within docker_client methods. Just that the parent for those contexts need to flow from here.

File: agent/eni/pause/pause_linux.go [hunk 15]
```diff
@@ -48,7 +49,9 @@ func loadFromFile(path string, dockerClient engine.DockerClient, fs os.FileSyste
 		return errors.Wrapf(err,
 			"pause container load: failed to read pause container image: %s", path)
 	}
-	if err := dockerClient.LoadImage(pauseContainerReader, engine.LoadImageTimeout); err != nil {
+	ctx, cancel := context.WithTimeout(context.TODO(), engine.LoadImageTimeout)
```
Comment (aaithal) [inline agent/eni/pause/pause_linux.go]: You shouldn't need to create a new context here, using `context.TODO()`. You should instead wire it from the "app" package (via loader. LoadImage)

File: agent/stats/engine.go [hunk 16]
```diff
@@ -123,7 +124,9 @@ func NewDockerStatsEngine(cfg *config.Config, client ecsengine.DockerClient, con
 
 // synchronizeState goes through all the containers on the instance to synchronize the state on agent start
 func (engine *DockerStatsEngine) synchronizeState() error {
-	listContainersResponse := engine.client.ListContainers(false, ecsengine.ListContainersTimeout)
+	ctx, cancel := context.WithTimeout(context.Background(), ecsengine.ListContainersTimeout)
```
Comment (aaithal) [inline agent/stats/engine.go]: Here too, you shouldn't need to create a new context. Just use `engine.ctx` as the parent and let the child context creation happen in `ListContainers`

Comment (haikuoliu) [reply agent/stats/engine.go]: There is no ctx attribute for stats engine, and I think creating a context for just one method seems to be redundant. So instead of creating a timeout context from background context, I will pass the background context directly to the method.

Comment (aaithal) [reply agent/stats/engine.go]: You can create one :) Let it be set from app.ctx.

Comment (haikuoliu) [reply agent/stats/engine.go]: ok, cool

File: agent/eni/pause/pause_linux.go (hunk 15 above)
Comment (haikuoliu) [reply agent/eni/pause/pause_linux.go]: ok, sounds good.

File: agent/engine/docker_image_manager.go (hunk 11 above)
Comment (haikuoliu) [reply agent/engine/docker_image_manager.go]: Previously, in the RemoveImage function of docker client, it uses context.Background() directly, 
so I use context.Background() in the caller, you want me to do the change?

File: agent/engine/docker_client.go [hunk 17]
```diff
@@ -164,9 +165,9 @@ type DockerClient interface {
 	InspectImage(string) (*docker.Image, error)
 
 	// RemoveImage removes the metadata associated with an image and may remove the underlying layer data. A timeout
-	// value should be provided for the request.
-	RemoveImage(string, time.Duration) error
-	LoadImage(io.Reader, time.Duration) error
+	// value and a context should be provided for the request.
+	RemoveImage(context.Context, string, time.Duration) error
+	LoadImage(context.Context, io.Reader, time.Duration) error
```
[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/engine/docker_client.go]: ok, added.

File: agent/engine/docker_client.go [hunk 18]
```diff
@@ -471,18 +472,13 @@ func (dg *dockerGoClient) getAuthdata(image string, authData *ecr.RegistryAuthen
 	return authConfig, nil
 }
 
-func (dg *dockerGoClient) CreateContainer(config *docker.Config,
+// The timeout parameter here is only used to assemble the return error message
```
[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/engine/docker_client.go]: ok, done.

File: agent/engine/docker_task_engine.go (hunk 12 above)
Comment (haikuoliu) [reply agent/engine/docker_task_engine.go]: Will change them all to engine.ctx.

File: agent/engine/docker_task_engine.go (hunk 13 above)
Comment (haikuoliu) [reply agent/engine/docker_task_engine.go]: See comments above.

File: agent/engine/docker_task_engine.go (hunk 14 above)
Comment (haikuoliu) [reply agent/engine/docker_task_engine.go]: See comments above.

File: agent/engine/docker_task_engine.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (haikuoliu) [reply agent/engine/docker_task_engine.go]: After the update for this PR is done, there is no need to pass the inspectContainerTimeout anymore.

File: agent/engine/docker_image_manager.go (hunk 11 above)
Comment (aaithal) [reply agent/engine/docker_image_manager.go]: yes please

**Comment by haikuoliu** [comment]
Update: add ctx attribute to the stats engine; move the creation of timeout context from caller to docker client API; wire context from caller instead of using context.TODO() or context.Background().

File: agent/engine/docker_client.go [hunk 19]
```diff
@@ -588,32 +582,28 @@ func dockerStateToState(state docker.State) api.ContainerStatus {
...
+	dockerContainer, err := dg.InspectContainer(ctx, dockerID, inspectContainerTimeout)
 	if err != nil {
 		return api.ContainerStatusNone, DockerContainerMetadata{Error: CannotDescribeContainerError{err}}
 	}
 	return dockerStateToState(dockerContainer.State), metadataFromContainer(dockerContainer)
 }
 
-func (dg *dockerGoClient) InspectContainer(dockerID string, timeout time.Duration) (*docker.Container, error) {
+// InspectContainer uses the timeout parameter only to assemble the return error message
+// TODO: replace the timeout with a string type error message
```
Comment (aaithal) [inline agent/engine/docker_client.go]: Can you get rid of this TODO (and others that were added)? It's out of date now.

File: agent/engine/docker_task_engine.go [hunk 20]
```diff
@@ -957,7 +955,7 @@ func (engine *DockerTaskEngine) buildCNIConfigFromTaskContainer(task *api.Task,
 	if !ok {
 		return nil, errors.New("engine: failed to find the pause container")
 	}
-	containerInspectOutput, err := engine.client.InspectContainer(pauseContainer.DockerName, inspectContainerTimeout)
+	containerInspectOutput, err := engine.client.InspectContainer(engine.ctx, pauseContainer.DockerName, inspectContainerTimeout)
```
Comment (aaithal) [inline agent/engine/docker_task_engine.go]: nit: this can be split into multiple lines to improve readability

File: agent/engine/docker_client_test.go [hunk 21]
```diff
@@ -363,28 +362,19 @@ func TestCreateContainerTimeout(t *testing.T) {
 	defer done()
 
 	warp := make(chan time.Time)
```
Comment (aaithal) [inline agent/engine/docker_client_test.go]: Is `warp` used/effective anymore? Can we get rid of that as well (here and in other methods)?

File: agent/engine/docker_client_test.go [hunk 22]
```diff
@@ -479,21 +468,17 @@ func TestStopContainerTimeout(t *testing.T) {
...
-	wait := &sync.WaitGroup{}
-	wait.Add(1)
 	mockDocker.EXPECT().StopContainerWithContext("id", uint(client.config.DockerStopTimeout/time.Second), gomock.Any()).Do(func(x, y, z interface{}) {
```
Comment (aaithal) [inline agent/engine/docker_client_test.go]: I think this needs `.MaxTimes(1)` as well. So do all other `EXPECT()`, where we are actually letting these timeouts time out (wherever `xContainerShortTimeout` is used).

File: agent/engine/docker_client_test.go [hunk 23]
```diff
@@ -363,28 +362,19 @@ func TestCreateContainerTimeout(t *testing.T) {
...
 	config := docker.CreateContainerOptions{Config: &docker.Config{Memory: 100}, Name: "containerName"}
 	mockDocker.EXPECT().CreateContainer(gomock.Any()).Do(func(x interface{}) {
 		warp <- time.Now()
-		wait.Wait()
-		// Don't return, verify timeout happens
-		// TODO remove the MaxTimes by cancel the context passed to CreateContainer
-		// when issue #1212 is resolved
 	}).MaxTimes(1)
-	metadata := client.CreateContainer(config.Config, nil, config.Name, xContainerShortTimeout)
+	metadata := client.CreateContainer(context.TODO(), config.Config, nil, config.Name, xContainerShortTimeout)
```
Comment (aaithal) [inline agent/engine/docker_client_test.go]: In all of the tests, where you have passed `context.TODO()` directly, please modify those to be something like this:

```golang
ctx, cancel := context.WithCancel(context.TODO())
defer cancel()

// pass ctx
```

Otherwise, we'll have go routine leaks because of uncancelled context in edge cases.

File: agent/engine/docker_image_manager_integ_test.go [hunk 24]
```diff
@@ -667,19 +668,19 @@ func verifyImagesAreNotRemoved(imageManager *dockerImageManager, imageIDs ...str
 }
 
 func cleanupImagesHappy(imageManager *dockerImageManager) {
-	imageManager.client.RemoveContainer("test1", removeContainerTimeout)
-	imageManager.client.RemoveContainer("test2", removeContainerTimeout)
-	imageManager.client.RemoveContainer("test3", removeContainerTimeout)
-	imageManager.client.RemoveImage(test1Image1Name, imageRemovalTimeout)
-	imageManager.client.RemoveImage(test1Image2Name, imageRemovalTimeout)
-	imageManager.client.RemoveImage(test1Image3Name, imageRemovalTimeout)
+	imageManager.client.RemoveContainer(context.TODO(), "test1", removeContainerTimeout)
```
Comment (aaithal) [inline agent/engine/docker_image_manager_integ_test.go]: Just create context once as per my earlier comments and use it everywhere.

File: agent/engine/docker_image_manager_integ_test.go [hunk 25]
```diff
@@ -667,19 +668,19 @@ func verifyImagesAreNotRemoved(imageManager *dockerImageManager, imageIDs ...str
...
 }
 
 func cleanupImagesThreshold(imageManager *dockerImageManager) {
-	imageManager.client.RemoveContainer("test1", removeContainerTimeout)
-	imageManager.client.RemoveContainer("test2", removeContainerTimeout)
-	imageManager.client.RemoveContainer("test3", removeContainerTimeout)
-	imageManager.client.RemoveImage(test2Image1Name, imageRemovalTimeout)
-	imageManager.client.RemoveImage(test2Image2Name, imageRemovalTimeout)
-	imageManager.client.RemoveImage(test2Image3Name, imageRemovalTimeout)
+	imageManager.client.RemoveContainer(context.TODO(), "test1", removeContainerTimeout)
```
Comment (aaithal) [inline agent/engine/docker_image_manager_integ_test.go]: Just create context once as per my earlier comments and use it everywhere.

File: agent/engine/docker_image_manager_test.go [hunk 26]
```diff
@@ -776,8 +776,8 @@ func TestImageCleanupCannotRemoveImage(t *testing.T) {
 	imageState.PulledAt = time.Now().AddDate(0, -2, 0)
 	imageState.LastUsedAt = time.Now().AddDate(0, -2, 0)
 
-	client.EXPECT().RemoveImage(container.Image, removeImageTimeout).Return(errors.New("error removing image")).AnyTimes()
-	imageManager.removeUnusedImages()
+	client.EXPECT().RemoveImage(gomock.Any(), container.Image, removeImageTimeout).Return(errors.New("error removing image")).AnyTimes()
+	imageManager.removeUnusedImages(context.TODO())
```
Comment (aaithal) [inline agent/engine/docker_image_manager_test.go]: Same comment as in the other file about using cancellable context everywhere in this file

File: agent/eni/pause/pause_linux_test.go [hunk 27]
```diff
@@ -52,7 +53,7 @@ func TestLoadFromFileWithReaderError(t *testing.T) {
 	mockfs := mock_os.NewMockFileSystem(ctrl)
 	mockfs.EXPECT().Open(pauseTarballPath).Return(nil, errors.New("Dummy Reader Error"))
 
-	err = loadFromFile(pauseTarballPath, client, mockfs)
+	err = loadFromFile(context.TODO(), pauseTarballPath, client, mockfs)
```
Comment (aaithal) [inline agent/eni/pause/pause_linux_test.go]: Same comment as in the other file about using cancellable context here

File: agent/stats/engine_integ_test.go [hunk 28]
```diff
@@ -86,7 +87,7 @@ func TestStatsEngineWithExistingContainersWithoutHealth(t *testing.T) {
 
 	// Simulate container start prior to listener initialization.
 	time.Sleep(checkPointSleep)
-	err = engine.MustInit(taskEngine, defaultCluster, defaultContainerInstance)
+	err = engine.MustInit(context.TODO(), taskEngine, defaultCluster, defaultContainerInstance)
```
Comment (aaithal) [inline agent/stats/engine_integ_test.go]: Same comment as in the other file about using cancellable context here

File: agent/stats/engine_test.go [hunk 29]
```diff
@@ -52,6 +52,7 @@ func TestStatsEngineAddRemoveContainers(t *testing.T) {
 	mockDockerClient.EXPECT().Stats(gomock.Any(), gomock.Any()).Return(mockStatsChannel, nil).AnyTimes()
 
 	engine := NewDockerStatsEngine(&cfg, nil, eventStream("TestStatsEngineAddRemoveContainers"))
+	engine.ctx = context.TODO()
```
Comment (aaithal) [inline agent/stats/engine_test.go]: Same comment as in the other file about using cancellable context here

File: agent/engine/docker_client.go (hunk 19 above)
Comment (haikuoliu) [reply agent/engine/docker_client.go]: This is the one that I added, we have talked about changing the timeout parameter into some kind of error message, and add a TODO for this purpose.

Comment (aaithal) [reply agent/engine/docker_client.go]: Right, that was when timeout was only being used for generating error messages. But, now its actually used for creating the context. So, that's not needed anymore

Comment (haikuoliu) [reply agent/engine/docker_client.go]: Oh I see, you are right!

File: agent/engine/docker_task_engine.go (hunk 20 above)
Comment (haikuoliu) [reply agent/engine/docker_task_engine.go]: OK

File: agent/engine/docker_client_test.go (hunk 21 above)
Comment (haikuoliu) [reply agent/engine/docker_client_test.go]: It's used to make the go routine not stop so that we don't have to handle the logic after the mocked function return. I will remove the warp add return value for the mocked function instead.

File: agent/engine/docker_client_test.go (hunk 22 above)
Comment (haikuoliu) [reply agent/engine/docker_client_test.go]: OK, added `.MaxTimes(1)` to all timeout tests in docker client.

File: agent/engine/docker_client_test.go (hunk 23 above)
Comment (haikuoliu) [reply agent/engine/docker_client_test.go]: OK, sure.

File: agent/engine/docker_image_manager_integ_test.go (hunk 25 above)
Comment (haikuoliu) [reply agent/engine/docker_image_manager_integ_test.go]: OK

File: agent/engine/docker_image_manager_integ_test.go (hunk 24 above)
Comment (haikuoliu) [reply agent/engine/docker_image_manager_integ_test.go]: OK

**Comment by haikuoliu** [comment]
Update: addressed all comments about the comments and tests.

File: agent/engine/docker_image_manager_test.go [hunk 30]
```diff
@@ -71,7 +71,9 @@ func TestImagePullRemoveDeadlock(t *testing.T) {
 	}()
 
 	go func() {
-		imageManager.(*dockerImageManager).removeUnusedImages()
+		ctx, cancel := context.WithCancel(context.TODO())
```
Comment (aaithal) [inline agent/engine/docker_image_manager_test.go]: can you move the context creation and `defer cancel()` methods outside of the go routine?

File: agent/engine/docker_image_manager_test.go [hunk 31]
```diff
@@ -1055,7 +1073,9 @@ func TestConcurrentRemoveUnusedImages(t *testing.T) {
 	for i := 0; i < numRoutines; i++ {
 		go func() {
 			<-ok
-			imageManager.removeUnusedImages()
+			ctx, cancel := context.WithCancel(context.TODO())
```
Comment (aaithal) [inline agent/engine/docker_image_manager_test.go]: can you move the context creation and `defer cancel()` methods outside of the `for` loop?

File: agent/stats/engine_integ_test.go [hunk 32]
```diff
@@ -143,7 +146,7 @@ func TestStatsEngineWithNewContainersWithoutHealth(t *testing.T) {
 		},
 		testTask)
 
-	err = engine.MustInit(taskEngine, defaultCluster, defaultContainerInstance)
+	err = engine.MustInit(context.TODO(), taskEngine, defaultCluster, defaultContainerInstance)
```
Comment (aaithal) [inline agent/stats/engine_integ_test.go]: You should use a cancellable context here similar to other test files (applies throughout this file)

File: agent/engine/docker_image_manager_test.go (hunk 30 above)
Comment (haikuoliu) [reply agent/engine/docker_image_manager_test.go]: OK

File: agent/engine/docker_image_manager_test.go (hunk 31 above)
Comment (haikuoliu) [reply agent/engine/docker_image_manager_test.go]: OK, here only one context should created and used.

File: agent/stats/engine_integ_test.go (hunk 32 above)
Comment (haikuoliu) [reply agent/stats/engine_integ_test.go]: ok, there are too many context creations in the test, I missed some, I will double check.

**Comment by haikuoliu** [comment]
Just pushed the commit, thank you guys!👍
